openshift/default-catalog-consistency/pkg/check/filesystem.go

```diff
@@ -38,3 +41,67 @@ func FilesystemHasDirectories(paths ...string) FilesystemCheck {
 		},
 	}
 }
+
+// FilesystemHasValidPathsOrLinks checks that the given paths exist inside the extracted image filesystem.
+// It supports regular files and symlinks:
+//   - If it's a regular file, we just check that it exists.
+//   - If it's a symlink, we check that the link target exists.
+//
+// This is useful for verifying container images where binaries may be symlinks,
+// In our image builds (see:
+// https://github.com/openshift/operator-framework-olm/blob/082d59a819afc43b24e9ca23c531bdfc35418722/operator-registry.Dockerfile#L16-L19),
+// the actual binaries are in /bin/registry/*, and /bin/* just has symlinks that point to them.
+//
+// This check avoids failing when the file is present but not marked executable —
+// for example, after image unpacking in some CI environments.
+func FilesystemHasValidPathsOrLinks(paths ...string) FilesystemCheck {
+	return FilesystemCheck{
+		Name: "FilesystemHasValidPathsOrLinks" + fmt.Sprintf(":(%q)", paths),
+
+		// We use the `os` package instead of fs.FS because fs.FS does not support Lstat or Readlink,
+		// which are needed to detect and resolve symlinks correctly.
+		// Therefore, we are looking to real file paths under the unpacked image (in tmpDir/fs).
+		Fn: func(_ context.Context, _ fs.FS, tmpDir string) error {
+			var errs []error
+			root := filepath.Join(tmpDir, "fs")
+
+			for _, rel := range paths {
+				fullPath := filepath.Join(root, rel)
+
+				fi, err := os.Lstat(fullPath)
+				if err != nil {
+					errs = append(errs, fmt.Errorf("error stating %q: %w", rel, err))
+					continue
+				}
+
+				if fi.Mode()&os.ModeSymlink != 0 {
+					target, err := os.Readlink(fullPath)
+					if err != nil {
+						errs = append(errs, fmt.Errorf("unreadable symlink %q: %w", rel, err))
+						continue
+					}
+
+					// Resolve the symlink target.
+					// Example 1: if the target is absolute like "/bin/registry/opm",
+					//   that means it should exist under the image root at tmpDir/fs/bin/registry/opm.
+					// Example 2: if the target is relative like "registry/opm" or "../registry/opm",
+					//   resolve it relative to the symlink’s own directory.
+					var resolved string
+					if filepath.IsAbs(target) {
+						// remove leading "/" and resolve from root
+						resolved = filepath.Join(root, target[1:])
+					} else {
+						// resolve relative path
+						resolved = filepath.Join(filepath.Dir(fullPath), target)
+					}
+
+					if _, err := os.Stat(resolved); err != nil {
+						errs = append(errs, fmt.Errorf("symlink %q points to missing target %q: %w", rel, resolved, err))
+					}
```

Comment:
@joelanford @bentito 


Sadly, my fist try to check if is an executable did not work out:

```
$ ls -la /var/folders/n4/j272tr6d7hq63mf7_skv6zr80000gn/T/oci-community-operator-index-2618243349/fs/bin/registry/opm
-rw-------  1 camilam  staff  91558832 20 May 10:30 /var/folders/n4/j272tr6d7hq63mf7_skv6zr80000gn/T/oci-community-operator-index-2618243349/fs/bin/registry/opm
```

When we unpack the image we lose the permissions bits, see: https://github.com/openshift/operator-framework-operator-controller/blob/main/openshift/default-catalog-consistency/pkg/extract/extract.go#L181-L185